api/oneinch/order.go

```diff
@@ -0,0 +1,131 @@
+package oneinch
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"net/http"
+
+	"github.com/svanas/1inch-sdk/golang/client/orderbook"
+	"github.com/svanas/1inch-sdk/golang/helpers/consts/contracts"
+	"github.com/svanas/ladder/api/web3"
+)
+
+func GetMakerAmount(order orderbook.OrderResponse) (*big.Int, error) {
+	i, ok := new(big.Int).SetString(order.Data.MakingAmount, 10)
+	if !ok {
+		return nil, fmt.Errorf("cannot convert %s to big.Int", order.Data.MakingAmount)
+	}
+	return i, nil
+}
+
+func GetTakerAmount(order orderbook.OrderResponse) (*big.Int, error) {
+	i, ok := new(big.Int).SetString(order.Data.TakingAmount, 10)
+	if !ok {
+		return nil, fmt.Errorf("cannot convert %s to big.Int", order.Data.TakingAmount)
+	}
+	return i, nil
+}
+
+func (client *Client) GetOrders(owner string) ([]orderbook.OrderResponse, error) {
+	oneInchClient, err := client.oneInchClient()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		page   float32 = 1
+		limit  float32 = 100
+		output []orderbook.OrderResponse
+	)
+	for {
+		orders, _, err := func() ([]orderbook.OrderResponse, *http.Response, error) {
+			beforeRequest()
+			defer afterRequest()
+			return oneInchClient.Orderbook.GetOrdersByCreatorAddress(context.Background(), orderbook.GetOrdersByCreatorAddressParams{
+				ChainId:        int(client.ChainId),
+				CreatorAddress: owner,
+				LimitOrderV3SubscribedApiControllerGetAllLimitOrdersParams: orderbook.LimitOrderV3SubscribedApiControllerGetAllLimitOrdersParams{
+					Page:   page,
+					Limit:  limit,
+					SortBy: "createDateTime",
+				},
+			})
+		}()
+		if err != nil {
+			return nil, err
+		}
+		output = append(output, orders...)
+		if len(orders) < int(limit) {
+			break
+		}
+		page++
+	}
+
+	return output, nil
+}
+
+func (client *Client) PlaceOrder(params orderbook.CreateOrderParams) error {
```

Comment:
The SDK's `CreateOrder` method will validate the params, use a Permit if possible, then default to Approvals if it needs to. The approval transaction is sent automatically if you pass in `SkipWanrings`, but gives the user a CLI prompt first if that config parameter is absent. If you want the user to *always* approve their token manually, I can add in another flag to enable that functionality too.

It also sends the request to the 1inch API so your bot doesn't need to manage the actual REST request side at all: https://github.com/1inch/1inch-sdk/blob/c9ee092295148aefe30dac273c1bb0e07cc63bee/golang/client/orderbook.go#L26-L26

Reply:
Correct, but we don't want/need that extra logic.

We surely don't want some dependency to prompt our users for interactive activity (rather than to return an error) and broadcast on-chain transactions to the network.

That is why we tap into [Client.Do](https://github.com/1inch/1inch-sdk/blob/c9ee092295148aefe30dac273c1bb0e07cc63bee/golang/client/client.go#L141) (rather than into [CreateOrder](https://github.com/1inch/1inch-sdk/blob/c9ee092295148aefe30dac273c1bb0e07cc63bee/golang/client/orderbook.go#L26))

Reply:
> we don't want/need that extra logic.
Ideally, the SDK is designed so you do want it! By default, it will try to use Permit1, then will fallback to approvals if needed. You can also set which default behavior you want with `ApprovalType`

> We surely don't want some dependency to prompt our users for interactive activity
That is why the `SkipWarnings` prompt exists. We might be changing this to invert the command flag so that no-prompt is the default, but I am still undecided on which is better.

> broadcast on-chain transactions to the network
I just added in a new flag you can set to return an error instead of broadcasting the approval transaction [here](https://github.com/1inch/1inch-sdk/pull/26/files). Just `FailIfApprovalIsNeeded` to `true` in the config param ([example](https://github.com/1inch/1inch-sdk/blob/217fbb8b921deed221e17194422da0da63fc2c5a/golang/client/orderbook_integration_test.go#L40))

Here is a rough PR showing that most of the logic you added to manually create the order can be simplified to a single call. I didn't test it locally, but it should get you 90% of the way to the simpler flow: https://github.com/svanas/ladder/pull/4/files